### PR TITLE
Supports DETR transformer in mmcv.

### PR DESCRIPTION
**Pull request by v-qjqs**
This PR implements the DETR transformer in mmcv.

### CODE REVIEW

File: mmcv/cnn/bricks/transformer.py [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+        out = self.attn(
+            query,
+            key,
+            value=value,
+            attn_mask=attn_mask,
+            key_padding_mask=key_padding_mask)[0]
+        return residual + self.dropout(out)
+
+    def __repr__(self):
+        repr_str = self.__class__.__name__
```
Comment (ZwwWayne) [inline mmcv/cnn/bricks/transformer.py:66]: docstring

File: mmcv/cnn/bricks/transformer.py [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+                 dropout=0.0):
+        super(FFN, self).__init__()
+        assert num_fcs >= 2, 'num_fcs should be no less ' \
+            f'than 2. got {num_fcs}.'
+        self.feat_channels = feat_channels
+        self.feedforward_channels = feedforward_channels
+        self.num_fcs = num_fcs
+        self.act_cfg = act_cfg
+        self.dropout = dropout
+        # NOTE only when act_cfg['type'] not in ['Tanh', 'PReLU',
```
Comment (ZwwWayne) [inline mmcv/cnn/bricks/transformer.py:99]: These comments should not be put here. The users should be aware of that in the documentation of `build_activation_layer` or some place else.

File: mmcv/cnn/bricks/transformer.py [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+
+    def forward(self, x, residual=None):
+        """Forward function for FFNs."""
+        if residual is None:
+            residual = x
+        x = self.layers(x)
+        return residual + self.dropout(x)
+
+    def __repr__(self):
+        repr_str = self.__class__.__name__
```
Comment (ZwwWayne) [inline mmcv/cnn/bricks/transformer.py:123]: docstring

File: mmcv/cnn/bricks/transformer.py [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+        self.num_layers = num_layers
+        self.feat_channels = feat_channels
+        self.num_heads = num_heads
+        self.feedforward_channels = feedforward_channels
+        self.dropout = dropout
+        self.order = order
+        self.act_cfg = act_cfg
+        self.norm_cfg = norm_cfg
+        self.num_fcs = num_fcs
+        self.normalize_before = order[0] == 'norm'
```
Comment (ZwwWayne) [inline mmcv/cnn/bricks/transformer.py:360]: normalize_before -> pre_norm

File: mmcv/cnn/bricks/transformer.py [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+    def forward(self,
+                x,
+                key=None,
+                value=None,
+                residual=None,
+                query_pos=None,
+                key_pos=None,
+                attn_mask=None,
+                key_padding_mask=None):
+        """Forward function for MultiheadAttention."""
```
Comment (xvjiarui) [inline mmcv/cnn/bricks/transformer.py:41]: This function is complicated and has many arguments. We may use a standard docstring.

File: mmcv/cnn/bricks/transformer.py [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+        norm_cfg (dict): Config dict for normalization layer. Default
+            layer normalization.
+        num_fcs (int): The number of fully-connected layers for FFNs.
+            Default 2.
+    """
+
+    def __init__(self,
+                 feat_channels,
+                 num_heads=8,
+                 feedforward_channels=2048,
```
Comment (xvjiarui) [inline mmcv/cnn/bricks/transformer.py:153]: We may leave these arguments unset.

File: mmcv/cnn/bricks/transformer.py [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+        for layer in self.order:
+            if layer == 'selfattn':
+                x = self.self_attn(
+                    x,
+                    x,
+                    x,
+                    inp_residual if self.normalize_before else None,
+                    query_pos,
+                    attn_mask=tgt_attn_mask,
+                    key_padding_mask=tgt_key_padding_mask)
```
Comment (xvjiarui) [inline mmcv/cnn/bricks/transformer.py:288]: Reformat this manually. This is not readable.

File: mmcv/cnn/bricks/transformer.py [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+        norm_cfg (dict): Same as `TransformerEncoderLayer`. Default
+            layer normalization.
+        num_fcs (int): Same as `TransformerEncoderLayer`. Default 2.
+    """
+
+    def __init__(self,
+                 num_layers,
+                 feat_channels,
+                 num_heads,
+                 feedforward_channels=2048,
```
Comment (xvjiarui) [inline mmcv/cnn/bricks/transformer.py:342]: We may leave this arg unset.

File: mmcv/cnn/bricks/transformer.py [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+        hs = self.decoder(
+            tgt,
+            memory,
+            memory_pos=pos_embed,
+            query_pos=query_embed,
+            memory_attn_mask=None,
+            tgt_attn_mask=None,
+            memory_key_padding_mask=mask,
+            tgt_key_padding_mask=None)
+        return hs.transpose(1, 2), memory.permute(1, 2, 0).view(bs, c, h, w)
```
Comment (xvjiarui) [inline mmcv/cnn/bricks/transformer.py:584]: `.reshape` instead of `.view` after `permute`.

File: mmcv/cnn/bricks/transformer.py [hunk 10]
```diff
@@ -0,0 +1,599 @@
...
+                 dropout=0.0,
+                 order=('selfattn', 'norm', 'multiheadattn', 'norm', 'ffn',
+                        'norm'),
+                 act_cfg=dict(type='ReLU', inplace=True),
+                 norm_cfg=dict(type='LN'),
+                 num_fcs=2):
+        super(TransformerDecoderLayer, self).__init__()
+        assert isinstance(order, tuple) and len(order) == 6
+        assert set(order) == set(
+            ['selfattn', 'norm', 'multiheadattn', 'norm', 'ffn', 'norm'])
```
Comment (xvjiarui) [inline mmcv/cnn/bricks/transformer.py:247]: Only one `'norm'` is enough.

File: mmcv/cnn/bricks/transformer.py [hunk 11]
```diff
@@ -0,0 +1,599 @@
...
+        repr_str += f'norm_cfg={self.norm_cfg}, '
+        repr_str += f'num_fcs={self.num_fcs})'
+        return repr_str
+
+
+class TransformerEncoder(nn.Module):
+    """Implements the encoder in transformer.
+
+    Args:
+        num_layers (int): The number of TransformerEncoderLayer.
```
Comment (xvjiarui) [inline mmcv/cnn/bricks/transformer.py:326]:
```suggestion
        num_layers (int): The number of `TransformerEncoderLayer`.
```

**Comment by xvjiarui** [comment]
For the complicated `forward` functions, we may use docstrings like torchvision.

File: mmcv/cnn/bricks/transformer.py [hunk 12]
```diff
@@ -0,0 +1,721 @@
...
+
+    def forward(self,
+                x,
+                memory,
+                memory_pos=None,
+                query_pos=None,
+                memory_attn_mask=None,
+                tgt_attn_mask=None,
+                memory_key_padding_mask=None,
+                tgt_key_padding_mask=None):
```
Comment (ZwwWayne) [inline mmcv/cnn/bricks/transformer.py:319]: tgt -> target

File: mmcv/cnn/bricks/transformer.py [hunk 13]
```diff
@@ -0,0 +1,721 @@
...
+                                        act_cfg, norm_cfg, num_fcs))
+        self.norm = build_norm_layer(norm_cfg, feat_channels)[1]
+
+    def forward(self,
+                x,
+                memory,
+                memory_pos=None,
+                query_pos=None,
+                memory_attn_mask=None,
+                tgt_attn_mask=None,
```
Comment (ZwwWayne) [inline mmcv/cnn/bricks/transformer.py:534]: tgt -> target

**Review by ZwwWayne — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

File: mmcv/cnn/bricks/transformer.py [hunk 14]
```diff
@@ -0,0 +1,721 @@
...
+    This module implements MultiheadAttention with residual connection.
+
+    Args:
+        feat_channels (int): The embedding dimension.
+        num_heads (int): Parallel attention heads. Same as
+            `nn.MultiheadAttention`.
+        dropout (float): A Dropout layer on attn_output_weights. Default 0.0.
+    """
+
+    def __init__(self, feat_channels, num_heads, dropout=0.0):
```
Comment (hellock) [inline mmcv/cnn/bricks/transformer.py:22]: Why `feat_channels` instead of `embed_dims`?

File: mmcv/cnn/bricks/transformer.py [hunk 15]
```diff
@@ -0,0 +1,721 @@
...
+from ..utils import xavier_init
+from .activation import build_activation_layer
+from .norm import build_norm_layer
+from .wrappers import Linear
+
+
+class MultiheadAttention(nn.Module):
+    """A warpper for torch.nn.MultiheadAttention.
+
+    This module implements MultiheadAttention with residual connection.
```
Comment (hellock) [inline mmcv/cnn/bricks/transformer.py:13]: Is this module general enough?

Comment (v-qjqs) [reply mmcv/cnn/bricks/transformer.py:13]: Hi, it implements self attention and multihead attention together, and plugins residual connection to format forwad_pre and forward_post in official DETR together.
